src/packet/outgoing.rs

```diff
@@ -0,0 +1,202 @@
+use crate::{
+    net::constants::{DEFAULT_ORDERING_STREAM, DEFAULT_SEQUENCING_STREAM},
+    packet::{
+        header::{
+            AckedPacketHeader, ArrangingHeader, FragmentHeader, HeaderWriter, StandardHeader,
+        },
+        DeliveryGuarantee, OrderingGuarantee, PacketType,
+    },
+};
+
+/// Builder that could be used to construct an outgoing laminar packet.
+pub struct OutgoingPacketBuilder<'p> {
```

Comment:
For this builder, I think it makes sense to store all of the intermediate bits on the builder and only parse to the Outgoing packet on build. This allows us to only have one allocation of a Vec when the outgoing packet is built.
This would make the signature of the `OutgoingPacket` be 
```
struct OutgoingPacket{
    contents: Vec<u8>
}
```

Reply:
I am not sure what you mean by ' intermediate bits'. I already have a reference to the payload in here. And it does already do one allocation over [here](https://github.com/amethyst/laminar/pull/150/files#diff-71b90166b67cde207975f7515922c988R118), when it is build.

Reply:
As you're adding bytes to the header `Vec`, the vector needs to grow (since you didn't initialize it with a capacity). Each time this vector grows (roughly doubling in size) it allocates a new underlying section of memory. Then, when you call `concat()` it does yet another allocation into a new `Vec` and returns it. My suggestion is to store all of the stack structures (the header structs) on the builder as defined, and only populate a new vector when the call to `build()` is made.

Reply:
@TimonPost I think this can be an optimization in a second round if you want to make an issue for it. This PR has already been outstanding for a long time, and is ginormous; I'd prefer to get this merged as it is becoming difficult to code around it.